### PR TITLE
capi: resolve imported globals

### PR DESCRIPTION
**Pull request by gumb0**
Depends on #637.

### CODE REVIEW

File: lib/fizzy/capi.cpp [hunk 1]
```diff
@@ -183,6 +183,31 @@ inline std::vector<fizzy::ImportedFunction> unwrap(
     return imported_functions;
 }
 
+inline fizzy::ImportedGlobal unwrap(const FizzyImportedGlobal& c_imported_global)
+{
+    fizzy::ImportedGlobal imported_global;
+    imported_global.module =
+        c_imported_global.module ? std::string{c_imported_global.module} : std::string{};
```
Comment (chfast) [inline lib/fizzy/capi.cpp:190]: The `c_imported_global.module` cannot be null.

**Comment by axic** [comment]
Is this a replacement of #637?

**Comment by gumb0** [comment]
> Is this a replacement of #637?

#637 is C++ part, and this one is C part on top of it.

File: include/fizzy/fizzy.h [hunk 2]
```diff
@@ -406,14 +417,13 @@ FizzyInstance* fizzy_instantiate(const FizzyModule* module,
 /// FizzyImportedFunction::name strings.
 ///
 /// @note
-/// Function expects @p imported_globals to be in the order of imports defined in the module.
-/// No validation is done on the number of globals passed in, nor on their order.
-/// When number of passed globals or their order is different from the one defined by the
-/// module, behaviour is undefined.
+/// Globals in @a imported_globals are allowed to be in any order and allowed to include some
```
Comment (axic) [inline include/fizzy/fizzy.h:420]: Are we using `#imported_globals` now? Or not for arguments?

Comment (gumb0) [reply include/fizzy/fizzy.h:420]: Not for arguments, I think I tried it already https://github.com/wasmx/fizzy/pull/681#issuecomment-762415754

Comment (axic) [inline include/fizzy/fizzy.h:420]: Does the of lifetime `module`/`name` needs to be only valid during `fizzy_resolve_instantiate`?

File: test/unittests/capi_test.cpp [hunk 3]
```diff
@@ -751,6 +751,118 @@ TEST(capi, resolve_instantiate_function_duplicate)
...
+
+    // reordered globals
+    module = fizzy_parse(wasm.data(), wasm.size());
+    ASSERT_NE(module, nullptr);
+    FizzyImportedGlobal host_globals_reordered[] = {{"mod1", "g2", mod1g2}, {"mod2", "g1", mod2g1},
+        {"mod2", "g2", mod2g2}, {"mod1", "g1", mod1g1}};
+    instance = fizzy_resolve_instantiate(
+        module, &mod1foo1, 1, nullptr, nullptr, host_globals_reordered, 4);
+    EXPECT_NE(instance, nullptr);
+    fizzy_free_instance(instance);
```
Comment (axic) [inline test/unittests/capi_test.cpp:833]: Just like in the C++ test, probably check the values here too.

File: test/unittests/capi_test.cpp [hunk 4]
```diff
@@ -751,6 +751,118 @@ TEST(capi, resolve_instantiate_function_duplicate)
...
+
+    // extra globals
+    module = fizzy_parse(wasm.data(), wasm.size());
+    ASSERT_NE(module, nullptr);
+    FizzyImportedGlobal host_globals_extra[] = {{"mod1", "g1", mod1g1}, {"mod1", "g2", mod1g2},
+        {"mod2", "g1", mod2g1}, {"mod2", "g2", mod2g2}, {"mod3", "g1", mod1g1}};
+    instance =
+        fizzy_resolve_instantiate(module, &mod1foo1, 1, nullptr, nullptr, host_globals_extra, 4);
+    EXPECT_NE(instance, nullptr);
+    fizzy_free_instance(instance);
```
Comment (axic) [inline test/unittests/capi_test.cpp:843]: And here.

File: test/unittests/capi_test.cpp [hunk 5]
```diff
@@ -667,34 +667,51 @@ TEST(capi, resolve_instantiate_functions)
...
     const FizzyValueType input_type = FizzyValueTypeI32;
-    FizzyValue result_int{42};
-    FizzyExternalFunction mod1foo1 = {{FizzyValueTypeI32, &input_type, 1}, NullFn, &result_int};
-    FizzyExternalFunction mod1foo2 = {{FizzyValueTypeI64, &input_type, 1}, NullFn, &result_int};
+    FizzyValue result_int32{42};
+    FizzyExternalFunction mod1foo1 = {{FizzyValueTypeI32, &input_type, 1}, host_fn, &result_int32};
+    FizzyValue result_int64{43};
+    FizzyExternalFunction mod1foo2 = {{FizzyValueTypeI64, &input_type, 1}, host_fn, &result_int64};
     FizzyValue result_f32;
-    result_f32.f32 = 42;
```
Comment (axic) [inline test/unittests/capi_test.cpp:687]: Ugh.

File: include/fizzy/fizzy.h (hunk 2 above)
Comment (gumb0) [reply include/fizzy/fizzy.h:420]: Added a note.

File: test/unittests/capi_test.cpp (hunk 4 above)
Comment (gumb0) [reply test/unittests/capi_test.cpp:855]: Added.

File: test/unittests/capi_test.cpp (hunk 3 above)
Comment (gumb0) [reply test/unittests/capi_test.cpp:839]: Added.

**Comment by axic** [comment]
Wasn't the last commit merged already?

**Comment by gumb0** [comment]
> Wasn't the last commit merged already?

It's base was this PR, so you merged it here.
(it requires `FizzyImportedGlobal`)

**Comment by axic** [comment]
Oh, sorry for that!

**Comment by axic** [comment]
Recreated it in #701.

**Review by axic — APPROVED** [review]

**Pull request merged**